blockstore/badger/blockstore.go

```diff
@@ -732,6 +732,20 @@ func (b *Blockstore) Put(ctx context.Context, block blocks.Block) error {
 	}
 
 	put := func(db *badger.DB) error {
+		// Check if we have it before writing it.
+		switch err := db.View(func(txn *badger.Txn) error {
+			_, err := txn.Get(k)
+			return err
+		}); err {
+		case badger.ErrKeyNotFound:
+		case nil:
+			// Already exists, skip the put.
+			return nil
+		default:
+			return err
+		}
+
+		// Then write it.
```

Comment:
Should be relatively more efficient to do this check inside `db.Update` below and discarding the transaction if the key exists, instead of starting a new transaction. Although in practice it probably doesn't matter much.